core/src/main/java/org/apache/calcite/rel/rules/AggregateProjectConstantToDummyJoinRule.java

```diff
@@ -0,0 +1,147 @@
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.core.Aggregate;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rel.core.Project;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexLiteral;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.util.ImmutableBitSet;
+
+import com.google.common.collect.ImmutableList;
+
+import org.immutables.value.Value;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Planner rule that recognizes  a {@link org.apache.calcite.rel.core.Aggregate}
+ * on top of a {@link org.apache.calcite.rel.core.Project} where the aggregate's group set
+ * contains boolean literals (true, false), and removes the literals from the group keys by joining
+ * with a dummy table of boolean literals.
+ *
+ * <pre>{@code
+ * select avg(sal)
+ * from emp
+ * group by true;
+ * }</pre>
+ * becomes
+ * <pre>{@code
+ * select avg(sal)
+ * from emp, (select true x) dummy
+ * group by dummy.x;
+ * }</pre>
+ */
+@Value.Enclosing
+public final class AggregateProjectConstantToDummyJoinRule
+    extends RelRule<AggregateProjectConstantToDummyJoinRule.Config> {
+
+  /** Creates an AggregateProjectConstantToDummyJoinRule. */
+  private AggregateProjectConstantToDummyJoinRule(Config config) {
+    super(config);
+  }
+
+  @Override public boolean matches(RelOptRuleCall call) {
+    final Aggregate aggregate = call.rel(0);
+    final Project project = call.rel(1);
+
+    for (int groupKey: aggregate.getGroupSet().asList()) {
+      if (groupKey >= aggregate.getRowType().getFieldCount()) {
+        continue;
+      }
+      RexNode groupKeyProject = project.getProjects().get(groupKey);
+      if (groupKeyProject instanceof RexLiteral) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Aggregate aggregate = call.rel(0);
+    final Project project = call.rel(1);
+
+    RelBuilder builder = call.builder();
+    RexBuilder rexBuilder = builder.getRexBuilder();
+
+    builder.push(project.getInput());
+    int offset = project.getInput().getRowType().getFieldCount();
+
+    final ImmutableList.Builder<ImmutableList<RexLiteral>> tuples =
+        ImmutableList.builder();
+    ImmutableList.Builder<RexLiteral> b = ImmutableList.builder();
+    RelDataTypeFactory.Builder relDataTypeBuilder = rexBuilder.getTypeFactory().builder();
+
+    List<RexNode> projects = project.getProjects();
+    for (int i = 0; i < projects.size(); i++) {
+      RexNode node = projects.get(i);
+      if (node instanceof RexLiteral) {
+        b.add((RexLiteral) node);
+        relDataTypeBuilder.add(project.getRowType().getFieldList().get(i));
+      }
+    }
+    tuples.add(b.build());
+    builder.values(tuples.build(), relDataTypeBuilder.build());
```

Comment:
```suggestion
    RelDataTypeFactory.Builder valuesType = rexBuilder.getTypeFactory().builder();
    List<RexLiteral> literals = new ArrayList<>();
    List<RexNode> projects = project.getProjects();
    for (int i = 0; i < projects.size(); i++) {
      RexNode node = projects.get(i);
      if (node instanceof RexLiteral) {
        literals.add((RexLiteral) node);
        valuesType.add(project.getRowType().getFieldList().get(i));
      }
    }
    builder.values(ImmutableList.of(literals), valuesType.build());
```
Small suggestion for readability.